File: src/utilities/ConfigLoader/ConfigLoader.cpp
#include "ConfigLoader.hpp"

namespace cl {
Config::Config(std::string filename) {
	this->n_pairs = 0;
	std::string buffer;
	std::string line;
	std::ifstream file;
	file.open(filename);
	if (!file.is_open()) {
		CL_Error ex(CLErrorCode::FAILED_2_OPEN);
		std::cerr << ex.what() << std::endl;
		throw ex;
	}
	while (std::getline(file, line)) {
		if (line.length() == 0) {
			continue;
		}
		if (line.at(0) == '#') {
			continue;
		}
		if (line.at(line.size() - 1) != ';') {
			if (line.at(line.size() - 1) == ',') {
				buffer += line;
				continue;
			}
			else {
				CL_Error ex(CLErrorCode::SEMICOLON);
				std::cerr << ex.what() << std::endl;
				throw ex;
			}

		}
		buffer += line;
	}
	file.close();
	buffer.erase(std::remove(buffer.begin(), buffer.end(), '\t'), buffer.end());
	std::vector<std::string> raw_entries;
	split_string(raw_entries, buffer, ';');
	std::vector<std::string> raw_keyvals;
	int entry_idx = 0;
	for (unsigned int i = 0; i < raw_entries.size(); i++) {
		if (raw_entries[i].length() == 0) {
			continue;
		}
		split_string(raw_keyvals, raw_entries[i], '=');
		if (raw_keyvals.size() != 2) {
			this->Config_free();
			CL_Error ex(CLErrorCode::EQUALS_SIGN);
			std::cerr << ex.what() << std::endl;
			throw ex;
		}
		remove_side_spaces(raw_keyvals[0]);
		remove_side_spaces(raw_keyvals[1]);
		Value* new_val = new Value(raw_keyvals[1]);

		// Search if key exists in dictionary
		std::map<std::string, int>::const_iterator it = this->keys_index.find(raw_keyvals[0]);
		if (it == this->keys_index.end()) {
			this->keys_index.insert(std::make_pair(raw_keyvals[0], entry_idx));
			this->values.push_back(new_val);
			entry_idx++;
			this->n_pairs++;
		}
		else {
			int idx = it->second;
			delete this->values[idx];
			this->values[idx] = new_val;
		}
		raw_keyvals.clear();
	}
}

Config::Config() {
	this->n_pairs = 0;
}

Config::~Config() {
	this->Config_free();
}

void Config::split_string(std::vector<std::string>& processed, std::string& buffer, char delimiter) {
	int base = 0;
	std::string split;
	for (unsigned int i = 0; i < buffer.length(); i++) {
		if (buffer.at(i) == delimiter) {
			split = buffer.substr(base, i - base);
			if (split.length() != 0) {
				processed.push_back(split);
			}
			base = i + 1;
		}
		else if (i == buffer.length() - 1) {
			split = buffer.substr(base, i + 1 - base);
			if (split.length() != 0) {
				processed.push_back(split);
			}
			base = i + 1;
		}
	}
}

void Config::remove_side_spaces(std::string& raw_str) {
	size_t left_idx = 0;
	size_t right_idx = 0;
	for (unsigned int i = 0; i < raw_str.length(); i++) {
		if (raw_str.at(i) != ' ') {
			left_idx = i;
			break;
		}
	}
	raw_str = raw_str.substr(left_idx, raw_str.length());
	for (long long i = raw_str.length() - 1; i >= 0; i--) {
		if (raw_str.at(i) != ' ') {
			(i == raw_str.length() - 1) ? right_idx = raw_str.length() : right_idx = i + 1;
			break;
		}
	}
	raw_str = raw_str.substr(0, right_idx);
}

void Config::Config_free() {
	for (size_t i = 0; i < this->values.size(); i++) {
		delete this->values[i];
	}
}

int Config::get_n_entries() {
	return this->n_pairs;
}

std::vector<std::string> Config::get_all_keys() {
	std::vector<std::string> ret;
	ret.resize(this->n_pairs);
	for (std::map<std::string, int>::iterator it = this->keys_index.begin(); it != this->keys_index.end(); it++) {
		ret[it->second] = it->first;
	}
	return ret;
}

Value* Config::get_value(std::string key) {
	std::map<std::string, int>::const_iterator it = this->keys_index.find(key);
	if (it == this->keys_index.end()) {
		CL_Error ex(CLErrorCode::KEY_NOT_FOUND);
		std::cerr << ex.what() << std::endl;
		throw ex;
	}
	return this->values[it->second];
}

std::vector<std::pair<std::string, Value*>> Config::get_all_key_values() {
	std::vector<std::pair<std::string, Value*>> ret;
	ret.resize(this->n_pairs);
	for (std::map<std::string, int>::iterator it = this->keys_index.begin(); it != this->keys_index.end(); it++) {
		ret[it->second] = std::make_pair(it->first, this->get_value(it->first));
	}
	return ret;
}

bool Config::key_exists(std::string key) {
	std::map<std::string, int>::const_iterator it = this->keys_index.find(key);
	if (it == this->keys_index.end()) {
		return false;
	}
	return true;
}

void Config::add_entry(std::string key, int value) {
	std::map<std::string, int>::const_iterator it = this->keys_index.find(key);
	if (it != this->keys_index.end()) {
		CL_Error ex(CLErrorCode::ADD_KEY_REPEAT);
		std::cerr << ex.what() << std::endl;
		throw ex;
	}
	Value* new_val = new Value(value);
	this->keys_index.insert(std::make_pair(key, this->n_pairs));
	this->values.push_back(new_val);
	this->n_pairs++;
}

void Config::add_entry(std::string key, double value) {
	std::map<std::string, int>::const_iterator it = this->keys_index.find(key);
	if (it != this->keys_index.end()) {
		CL_Error ex(CLErrorCode::ADD_KEY_REPEAT);
		std::cerr << ex.what() << std::endl;
		throw ex;
	}
	Value* new_val = new Value(value);
	this->keys_index.insert(std::make_pair(key, this->n_pairs));
	this->values.push_back(new_val);
	this->n_pairs++;
}

void Config::add_entry(std::string key, std::string value) {
	std::map<std::string, int>::const_iterator it = this->keys_index.find(key);
	if (it != this->keys_index.end()) {
		CL_Error ex(CLErrorCode::ADD_KEY_REPEAT);
		std::cerr << ex.what() << std::endl;
		throw ex;
	}
	Value* new_val = new Value(value);
	this->keys_index.insert(std::make_pair(key, this->n_pairs));
	this->values.push_back(new_val);
	this->n_pairs++;
}

void Config::add_entry(std::string key, const char* value) {
	std::map<std::string, int>::const_iterator it = this->keys_index.find(key);
	if (it != this->keys_index.end()) {
		CL_Error ex(CLErrorCode::ADD_KEY_REPEAT);
		std::cerr << ex.what() << std::endl;
		throw ex;
	}
	Value* new_val = new Value(value);
	this->keys_index.insert(std::make_pair(key, this->n_pairs));
	this->values.push_back(new_val);
	this->n_pairs++;
}

void Config::add_entry(std::string key, List value) {
	std::map<std::string, int>::const_iterator it = this->keys_index.find(key);
	if (it != this->keys_index.end()) {
		CL_Error ex(CLErrorCode::ADD_KEY_REPEAT);
		std::cerr << ex.what() << std::endl;
		throw ex;
	}
	Value* new_val = new Value(value);
	this->keys_index.insert(std::make_pair(key, this->n_pairs));
	this->values.push_back(new_val);
	this->n_pairs++;
}

void Config::edit_value(std::string key, int value) {
	std::map<std::string, int>::const_iterator it = this->keys_index.find(key);
	if (it == this->keys_index.end()) {
		CL_Error ex(CLErrorCode::KEY_NOT_FOUND);
		std::cerr << ex.what() << std::endl;
		throw ex;
	}
	Value* new_val = new Value(value);
	int idx = it->second;
	delete this->values[idx];
	this->values[idx] = new_val;
}

void Config::edit_value(std::string key, double value) {
	std::map<std::string, int>::const_iterator it = this->keys_index.find(key);
	if (it == this->keys_index.end()) {
		CL_Error ex(CLErrorCode::KEY_NOT_FOUND);
		std::cerr << ex.what() << std::endl;
		throw ex;
	}
	Value* new_val = new Value(value);
	int idx = it->second;
	delete this->values[idx];
	this->values[idx] = new_val;
}

void Config::edit_value(std::string key, std::string value) {
	std::map<std::string, int>::const_iterator it = this->keys_index.find(key);
	if (it == this->keys_index.end()) {
		CL_Error ex(CLErrorCode::KEY_NOT_FOUND);
		std::cerr << ex.what() << std::endl;
		throw ex;
	}
	Value* new_val = new Value(value);
	int idx = it->second;
	delete this->values[idx];
	this->values[idx] = new_val;
}

void Config::edit_value(std::string key, const char* value) {
	std::map<std::string, int>::const_iterator it = this->keys_index.find(key);
	if (it == this->keys_index.end()) {
		CL_Error ex(CLErrorCode::KEY_NOT_FOUND);
		std::cerr << ex.what() << std::endl;
		throw ex;
	}
	Value* new_val = new Value(value);
	int idx = it->second;
	delete this->values[idx];
	this->values[idx] = new_val;
}

void Config::edit_value(std::string key, List value) {
	std::map<std::string, int>::const_iterator it = this->keys_index.find(key);
	if (it == this->keys_index.end()) {
		CL_Error ex(CLErrorCode::KEY_NOT_FOUND);
		std::cerr << ex.what() << std::endl;
		throw ex;
	}
	Value* new_val = new Value(value);
	int idx = it->second;
	delete this->values[idx];
	this->values[idx] = new_val;
}

void Config::save_config(std::string path) {
	std::ofstream out(path);

	std::string key;
	int value_idx;
	Value value;
	ValueType value_type;

	cl::List list_value;
	DataType list_value_type;
	for (std::map<std::string, int>::iterator it = this->keys_index.begin(); it != this->keys_index.end(); it++) {
		key = it->first;
		value_idx = it->second;
		value = *this->values[value_idx];
		value_type = value.get_type();
		out << key << " = ";
		// Value is either an integer, double, string, or list
		switch (value_type) {
		case ValueType::INT_NUM:
			out << std::to_string(value.get_data<int>()) << ";\n";
			break;
		case ValueType::DBL_NUM:
			out << std::to_string(value.get_data<double>()) << ";\n";
			break;
		case ValueType::STRING:
			out << value.get_data<std::string>() << ";\n";
			break;
		case ValueType::LIST:
			// If value is list, get integer, double, or string data from Variant object
			out << "[";
			list_value = value.get_data<cl::List>();
			for (int i = 0; i < list_value.size(); i++) {
				list_value_type = list_value[i].get_type();
				switch (list_value_type) {
				case DataType::INTEGER:
					out << std::to_string(list_value[i].get_data<int>());
					break;
				case DataType::DOUBLE:
					out << std::to_string(list_value[i].get_data<double>());
					break;
				case DataType::STRING:
					out << list_value[i].get_data<std::string>();
					break;
				case DataType::BOOLEAN:
				case DataType::CHARACTER:
				case DataType::UNDEFINED:
				default:
					out << "0";
					break;
				}
				if (i != list_value.size() - 1) {
					out << ", ";
				}
			}
			out << "]" << ";\n";
			break;
		case ValueType::UNDEF:
		default:
			out << "0";
			break;
		}
	}
	out.close();
}
}

File: src/utilities/ConfigLoader/ConfigLoader.py
import enum
import os.path

class CL_ErrorCodes(enum.Enum):
    OK = 0
    SEMICOLON = 1
    EQUALS_SIGN = 2
    KEY_NOT_FOUND = 3
    ADD_KEY_REPEAT = 4
    FAILED_2_OPEN = 5

class CL_Error(Exception):
    def __init__(self, message, error_code):
        super().__init__(message)
        self.err_code = error_code

class CL_ErrorWrapper:
    def __init__(self, error_code):
        self.error_code = error_code
        if(self.error_code == CL_ErrorCodes.OK):
            self.message = ''
        elif(self.error_code == CL_ErrorCodes.SEMICOLON):
            self.message = 'Missing semicolon in config file.'
        elif(self.error_code == CL_ErrorCodes.EQUALS_SIGN):
            self.message = 'Missing equals sign at an entry in config file.'
        elif(self.error_code == CL_ErrorCodes.KEY_NOT_FOUND):
            self.message = 'Key not found in config file.'
        elif(self.error_code == CL_ErrorCodes.ADD_KEY_REPEAT):
            self.message = 'Key already exists in config file.'
        elif(self.error_code == CL_ErrorCodes.FAILED_2_OPEN):
            self.message = 'Could not open/find specified config file.'
        else:
            self.message = 'Undefinded error.'

class Config:        
    def __init__(self, path=''):
        if(path == ''):
            self.contents = {}
            self.entries = 0
        else:
            self.contents = {}
            self.entries = 0
            buffer = ''
            if(not os.path.isfile(path)):
                ex = CL_ErrorWrapper(CL_ErrorCodes.FAILED_2_OPEN)
                raise CL_Error(ex.message, ex.error_code)
            with open(path) as file:
                for line in file:
                    line = line.replace('\n', '')
                    if(len(line) == 0):
                        continue
                    if(line[0] == '#'):
                        continue
                    if(line[len(line) - 1] != ';'):
                        if(line[len(line) - 1] == ','):
                            buffer += line
                            continue
                        else:
                            ex = CL_ErrorWrapper(CL_ErrorCodes.SEMICOLON)
                            raise CL_Error(ex.message, ex.error_code)
                    buffer += line
            buffer = buffer.replace('\t', '')
            split_buffer = buffer.split(';')
            for entry in split_buffer:
                if(len(entry) == 0):
                    continue
                key_val = entry.split('=')
                if(len(key_val) != 2):
                    ex = CL_ErrorWrapper(CL_ErrorCodes.EQUALS_SIGN)
                    raise CL_Error(ex.message, ex.error_code)
                key = self.__remove_side_spaces(key_val[0])
                value = self.__process_value(key_val[1])
                if(key in self.contents):
                    self.contents[key] = value
                else:    
                    self.contents[key] = value
                    self.entries += 1

    def __remove_side_spaces(self, raw_string):
        left_idx = 0
        right_idx = 0
        for i in range(0, len(raw_string)):
            if(raw_string[i] != ' '):
                left_idx = i
                break
        raw_string = raw_string[left_idx:]
        for i in range(len(raw_string) - 1, -1, -1):
            if(raw_string[i] != ' '):
                if(i == len(raw_string) - 1):
                    right_idx = len(raw_string)
                else:
                    right_idx = i + 1
                break
        raw_string = raw_string[:right_idx]
        return raw_string

    def __process_value(self, raw_string):
        raw_string = self.__remove_side_spaces(raw_string)
        if(raw_string[0] == '[' and raw_string[len(raw_string) - 1] == ']'):
            raw_string = raw_string[1:len(raw_string)-1]
            split_list = raw_string.split(',')
            for i in range(0, len(split_list)):
                split_list[i] = self.__remove_side_spaces(split_list[i])
                try:
                    to_float = float(split_list[i])
                    split_list[i] = to_float
                except:
                    continue
            return split_list
        else:
            raw_string = self.__remove_side_spaces(raw_string)
            try:
                to_float = float(raw_string)
                return to_float
            except:
                return raw_string

    def get_n_entries(self):
        return self.entries

    def get_all_keys(self):
        return [i for i in self.contents.keys()]

    def get_value(self, key):
        if(key not in self.contents):
            ex = CL_ErrorWrapper(CL_ErrorCodes.KEY_NOT_FOUND)
            raise CL_Error(ex.message, ex.error_code)
        return self.contents[key]

    def get_all_key_values(self):
        key_vals = []
        keys = [i for i in self.contents.keys()]
        for i in range(0, self.entries):
            key_vals.append((keys[i], self.get_value(keys[i])))
        return key_vals

    def add_entry(self, key, value):
        if(key in self.contents):
            ex = CL_ErrorWrapper(CL_ErrorCodes.ADD_KEY_REPEAT)
            raise CL_Error(ex.message, ex.error_code)
        self.contents[key] = value
        self.entries += 1

    def edit_value(self, key, value):
        if(key not in self.contents):
            ex = CL_ErrorWrapper(CL_ErrorCodes.KEY_NOT_FOUND)
            raise CL_Error(ex.message, ex.error_code)
        self.contents[key] = value

    def save_config(self, path):
        with open(path, 'w') as file:
            keys = [i for i in self.contents.keys()]
            for i in range(0, self.entries):
                file.write(keys[i] + ' = ')
                file.write(str(self.contents[keys[i]]) + ';\n')
